src/fft.rs

```diff
@@ -0,0 +1,200 @@
+use crate::util::parallelize;
+use ff::{Field, PrimeField};
+use group::{GroupOpsOwned, ScalarMulOwned};
+pub use halo2curves::{CurveAffine, CurveExt};
+
+/// Given FFT domain size k, return the omega in case of fft
+/// or return the omega_inv in case if ifft
+pub(crate) fn get_omega_or_inv<F: PrimeField>(k: u32, is_inverse: bool) -> F {
+    let mut omega_or_inv = if is_inverse {
+        F::ROOT_OF_UNITY_INV
+    } else {
+        F::ROOT_OF_UNITY
+    };
+    for _ in k..F::S {
+        omega_or_inv = omega_or_inv.square();
+    }
+    omega_or_inv
+}
+
+pub(crate) fn get_ifft_divisor<F: PrimeField>(k: u32) -> F {
+    F::TWO_INV.pow_vartime([k as u64, 0, 0, 0])
```

Comment:
What is the tail of zeroes for? It sort of causes it to find 64*3 times `ONE.square()` inside and only then calculates the power

Reply:
same here in case you need different size ntts might be better to keep those calculated values for intt normalization

Reply:
Frankly speaking, I don't understand your thesis, at the moment it just performs unnecessary operations, namely 192 times calculates the root of one

Reply:
Changed. As background: it is a normal representation for using 4 u64 to represent a field element. However, in our case, k will not that large.